opentelemetry-otlp/tests/integration_test/src/asserter.rs

```diff
@@ -0,0 +1,50 @@
+use opentelemetry_proto::tonic::trace::v1::ResourceSpans;
+use std::fs::File;
+
+// assert two resource spans is the same except for the timestamps in it
```

Comment:
we can do similar exception for spanid/traceid as well, to keep this simpler.

Reply:
I think for span ID and trace ID I'd rather use a `IncrementIdGenerator` so that we get predictable IDs. Because we want to make sure the parent-child relationship is correct

Reply:
In my view, that is partially defeating the purpose of integration tests.
Integration tests should be executing same code as a user would be doing, and custom IdGenerator deviates from that. 
Testing parent-child relation is not the goal of integration tests, right? 

Here's what I am thinking:
1. Execute the same hello world style example code + OTLP Exporter
2. Setup OTel Collector and have it save telemetry to file.
3. Validate that the file has expected telemetry. Things like attributes of diff. types are received, span events are received, metrics are correct etc. - You can of course validate that span's parent-child relation is maintained without pre-knowing the actual traceid/spanid.

As a first step, I'd suggest to start with simply validating that a file is created by collector, and we can add validation of its content as a next PR, to keep things moving easily.

Reply:
Yeah the current setup complete the 1 and 2. 

> Integration tests should be executing same code as a user would be doing, and custom IdGenerator deviates from that.

I think the `IdGenerator` is a small enough component that self-contained and can be well tested so the chance we missed anything bu not using the `RandomIdGenerator` is minimal

> You can of course validate that span's parent-child relation is maintained without pre-knowing the actual traceid/spanid.

Yeah it's boils down to a compare two forests. But I didn't find a easy way to write the expected span tress before asserting. I think a prediable ID is a good start point for now until we start testing context progatation through multiple containers. 

In other words, if we are only testing export from single container to collector. `IncrementIdGenerator` can satisify the requirment to be unique in this simple system. Randomness is not required by spec(though it's preferred)

Reply:
> Yeah the current setup complete the 1 and 2.


Not quite as this is executing diff. code than user would be (due to idgenerator). IMHO, Intergration tests should treat the library as blackbox only, and dont change anything at all, including idgenerator.

Reply:
Not sure I understand. 

> the current setup complete the 1 and 2.

The current setup includes the following steps:
- Execute the same hello world style example code + OTLP Exporter. This is done by spin up a collector container using testcontainer create and mount a local file to collector container.
- Setup OTel Collector and have it save telemetry to file. The result will be written to the mounted file by collector container

I think the difference is on how do we assert the result. The current setup opt in a quick appoaches where we parse the json output from collector and examine the `ResourceSpan` is expected, which need the predictable trace IDs and span IDs. 

I agree on the long term we should take a step further and rebuild the span relationship from `ResourceSpan`s but not sure if it should be a blocker for this PR

Reply:
> examine the ResourceSpan is expected, which need the predictable trace IDs and span IDs.

Why need predictable IDs? We don't use predictable timestamps, right? I was asking to do same for traceid,spanid as well. Anything that is auto-generated (and hence unpredictable), can be avoided from the assertion. Anything user-provided, and hence predictable can be validated.

To reiterate my opinion:
We should not use custom id generator, and instead only execute test app as a normal user would execute. I understand IdGenerator is a small change only, but that is still partially defeating the purpose of integration tests with otlp!

Reply:
>  I was asking to do same for traceid,spanid as well

I think the integration tests don't add much values if we cannot validate parent-children relationship between spans thus I am hesitiate to just ignore the trace IDs and span IDs. For example, say we have a bug that caused every span to be a root span. Without assertion on span IDs and traces IDs the integration test won't catch this. 

> Why need predictable IDs? We don't use predictable timestamps, right?

One point is we do allow users to provide their own `IdGenerator` while we don't allow users to bring their own clock

Either case if you feel strong about the predictable IdGenerator I can explore other options to assert the span relationships 🙂

Reply:
> I think the integration tests don't add much values if we cannot validate parent-children relationship

Agree. But we dont need predictable ids for that. We just need to validate that for the root span, `parentSpanId` will be empty, and the `parentSpanId` of the child will be the `spanId` of the parent. The values themselves can be anything! Its totally okay to add this validation in the future.

Reply:
> We just need to validate that for the root span, parentSpanId will be empty, and the parentSpanId of the child will be the spanId of the parent

This only validate the number of root spans and non root spans. To fully validate the span relationship you need to build two span forests and build a map between each spans in expected result and actual results, then validate they are the same. We will also need to handle links, which make the question even harder that we need build and compare two directed cyclic graph.

Prediable ID avoid all those operation with small drawback that we are not testing the random generation part of the operation, which IMO is a good trade off.

Reply:
_but we did seen a lot of issues with context propagation_

That should be tested via InMemoryExporter, right? Why its part of integration tests ?